[FSDP] Fix `use_orig_params=True` + `no_sync()`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #90548 [FSDP][Easy] ufmt files
* **#90546 [FSDP] Fix `use_orig_params=True` + `no_sync()`**

`no_sync()` introduces a separate case where a `FlatParameter` maintains an _unsharded_ gradient, instead of a _sharded_ one. This PR fixes `no_sync()` with `use_orig_params=True` by dealing with this separate case.

The existing `use_orig_params=False` already bypasses the built-in parameter/gradient size check, where the `flat_param` is sharded, while the `flat_param.grad` is unsharded. For `use_orig_params=True`, we need to use the same `.data` hack to side step the size check that we used to side step the dtype check for `keep_low_precision_grads=True`.

Commit:
[FSDP] Fix `use_orig_params=True` + `no_sync()`

[ghstack-poisoned]

Commit:
Update on "[FSDP] Fix `use_orig_params=True` + `no_sync()`"


`no_sync()` introduces a separate case where a `FlatParameter` maintains an _unsharded_ gradient, instead of a _sharded_ one. This PR fixes `no_sync()` with `use_orig_params=True` by dealing with this separate case.

The existing `use_orig_params=False` already bypasses the built-in parameter/gradient size check, where the `flat_param` is sharded, while the `flat_param.grad` is unsharded. For `use_orig_params=True`, we need to use the same `.data` hack to side step the size check that we used to side step the dtype check for `keep_low_precision_grads = True`.



[ghstack-poisoned]

Commit:
Update on "[FSDP] Fix `use_orig_params=True` + `no_sync()`"


`no_sync()` introduces a separate case where a `FlatParameter` maintains an _unsharded_ gradient, instead of a _sharded_ one. This PR fixes `no_sync()` with `use_orig_params=True` by dealing with this separate case.

The existing `use_orig_params=False` already bypasses the built-in parameter/gradient size check, where the `flat_param` is sharded, while the `flat_param.grad` is unsharded. For `use_orig_params=True`, we need to use the same `.data` hack to side step the size check that we used to side step the dtype check for `keep_low_precision_grads=True`.



[ghstack-poisoned]

Commit:
Update on "[FSDP] Fix `use_orig_params=True` + `no_sync()`"


`no_sync()` introduces a separate case where a `FlatParameter` maintains an _unsharded_ gradient, instead of a _sharded_ one. This PR fixes `no_sync()` with `use_orig_params=True` by dealing with this separate case.

The existing `use_orig_params=False` already bypasses the built-in parameter/gradient size check, where the `flat_param` is sharded, while the `flat_param.grad` is unsharded. For `use_orig_params=True`, we need to use the same `.data` hack to side step the size check that we used to side step the dtype check for `keep_low_precision_grads=True`.



[ghstack-poisoned]

mrshenli commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -1398,7 +1409,16 @@ def _use_unsharded_grad_views(self) -> None:
+                # `param` has the sharded size; `grad` has the unsharded size.
+                # This happens when running in `no_sync()`.
+                if param.grad is None:
+                    param.grad = torch.empty_like(param)
```
I assume this gonna fail if we create a zero-sized tensor here? And this is what you referred to as "the check that parameter/gradient sizes match"?

awgu replied:
We should not ever create a zero-sized tensor here because this is `_use_unsharded_grad_views()`, not `_use_sharded_grad_views()`.

The check I am referring to is that for any `param: nn.Parameter`, PyTorch checks that `param` and `param.grad` have the same shape if `param.grad is not None`. However, here, `param` may have the sharded size (e.g. `[5]`), while `param.grad may have the unsharded size (e.g.`[9]`). (The example numbers here show the padding case for 2 ranks and an unsharded size of 10 numel.)

mrshenli commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -1398,7 +1409,16 @@ def _use_unsharded_grad_views(self) -> None:
             )
             param = getattr(module, param_name)
-            param.grad = view
+            if param.shape != view.shape:
```
I assume this should only happen under `no_sync` mode. Do we need to assert that?

awgu replied:
Yes, ideally we can assert that, but at the moment, we do not propagate the `state._sync_gradients` attribute to the `FlatParamHandle`. When I wrote this PR a few days ago, I was not confident whether we should do that.

I like the design where the `state` _orchestrates_ the `FlatParamHandle`s. It becomes tricky when we have to maintain coherence between the `state` and `handles`. Currently, we do this for some things like mixed precision, and it causes some problems (as evidenced by my other PR stack).

mrshenli commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -1413,7 +1433,14 @@ def _use_unsharded_grad_views(self) -> None:
             param = getattr(module, param_name)
             prim_param = getattr(prim_module, prim_param_name)
-            param.grad = prim_param.grad
+            if param.shape != prim_param.grad.shape:
```
just for my own education, where can I learn more about why do we need to set grad in the `if` clause. I assume we are not going to accumulate grad to sharded param gradients?

awgu replied:
For this `if`, the "primary parameter" `prim_param` must have `.grad` set from the previous `for` loop in this method.

Now, this `for` loop is over the shared parameters, and this `if` is just checking the same shape mismatch case as before (i.e. `param` is sharded but its gradient is unsharded).

We should still set the `param.grad` to be `prim_param.grad`, which is a view into the unsharded gradient because for `no_sync()`, we may accumulate an _unsharded_ gradient.

The invariant of `use_orig_params=True` path is that each original parameter and its gradient should always be a view into the underlying `flat_param` and `flat_param.grad`, respectively. Whether the user actually cares that we set `param.grad` to be an unsharded view into `flat_param.grad` during `no_sync()` is debatable, but this is an important design choice that makes reasoning about the system more tractable (at least to me). This is why I was a big proponent of the `_use_sharded_views`, `_use_sharded_grad_views`, `_use_unsharded_views`, `_use_unsharded_grad_views` approach, where any change to the `flat_param` storage should include calls to these methods.

rohan-varma commented on test/distributed/fsdp/test_fsdp_grad_acc.py:
```diff
@@ -255,10 +269,12 @@ def _get_subtest_config(self) -> Dict[str, List[Any]]:
             ShardingStrategy.NO_SHARD,
         ],
     )
+    @parametrize("use_orig_params", [True])
```
what about use_orig_params=False?

awgu replied:
Apologies for the window of no signal for this test.

awgu replied:
Just kidding, the merge failed because this PR is too stale. I will fix in this PR.

rohan-varma commented on test/distributed/fsdp/test_fsdp_use_orig_params.py:
```diff
@@ -1048,10 +1049,122 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
+        )
+        test_model = FSDP(model, use_orig_params=True, **fsdp_kwargs)
+        baseline_optim = torch.optim.AdamW(baseline_model.parameters(), foreach=True)
+        test_optim = torch.optim.AdamW(test_model.parameters(), foreach=True)
```
nit: might be better to just name these model_no_use_orig and model_use_orig so it's clearer which one is which and what we are comparing.

awgu replied:
Changed.

rohan-varma commented on test/distributed/fsdp/test_fsdp_use_orig_params.py:
```diff
@@ -1048,10 +1049,122 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
+        # Compute some reference gradients using one forward/backward
+        baseline_out = baseline_model(inp)
+        test_out = test_model(inp)
+        torch.testing.assert_close(baseline_out, test_out)
```
Is this more of a relaxed assert than `self.assertEqual`? When should we use one over the other?

awgu replied:
Good question. Let me investigate.

rohan-varma commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -1247,10 +1247,21 @@ def _use_sharded_flat_param(self) -> None:
             # For the post-forward reshard, we may try to use sharded gradient
-            # views, but for the post-backward reshard, we delay the call to
-            # after the reduce-scatter
+            # views (or unsharded gradient views if a gradient was accumulated
```
just for my understanding, is this because we want the grad views to be unsharded since we were in no_sync, we didn't actually reduce_scatter and reshard?

awgu replied:
Yup

rohan-varma reviewed (approved):
LGTM, thanks for the thorough analysis and fix!

awgu commented on test/distributed/fsdp/test_fsdp_grad_acc.py:
```diff
@@ -128,11 +133,12 @@ def _test_grad_acc(
                 self.process_group,
                 FSDPInitMode.RECURSIVE,
-                CUDAInitMode.CUDA_AFTER,
+                CUDAInitMode.CUDA_BEFORE,
```
This was just to silence the warning about having flattening/sharding happening on CPU. There is no reason it needs to be `CUDA_BEFORE`. I am not sure why I did that in the first place.

awgu commented on test/distributed/fsdp/test_fsdp_grad_acc.py:
```diff
@@ -197,7 +207,11 @@ def permute_tensor(x: torch.Tensor):
+            acc_grads = [
+                p.grad.detach().clone()
+                for p in fsdp_model.parameters()
+                if p.grad is not None
```
`use_orig_params=True` will always require this idiom (for better or for worse) because even though an original parameter requires gradient, it may not receive a gradient if it is not in the local shard. I think this design matches the SPMD spirit best, but it just requires always having this `if grad is not None`.

awgu commented:
@pytorchbot merge

Commit:
Update on "[FSDP] Fix `use_orig_params=True` + `no_sync()`"


`no_sync()` introduces a separate case where a `FlatParameter` maintains an _unsharded_ gradient, instead of a _sharded_ one. This PR fixes `no_sync()` with `use_orig_params=True` by dealing with this separate case.

The existing `use_orig_params=False` already bypasses the built-in parameter/gradient size check, where the `flat_param` is sharded, while the `flat_param.grad` is unsharded. For `use_orig_params=True`, we need to use the same `.data` hack to side step the size check that we used to side step the dtype check for `keep_low_precision_grads=True`.



[ghstack-poisoned]

awgu commented:
@pytorchbot merge
